### ISSUE TITLE
[SE] Fault-tolernace and Checkpointing - Master Task

### ISSUE THREAD

**Issue by anovv** [volga-project/volga#3]
This is the main component for streaming fault tolerance. We should implement Asynchronous Barrier Snapshotting (Async Chandy–Lamport algorithm) with distributed snapshots and a proper state backend abstraction. There is a lot of info on the Internet on how Flink implements this. In short, we will need:

1. A common interface/serializable data model to maintain an Operator state and in-flight messages leading up to the current state in the checkpoint time window (similar to event sourcing)
2. A state backend interface to store/restore Operators state + messages and connectors to write/read existing storages (S3, DynamoDB, etc.)
3. Async Chandy–Lamport algo relies on checkpoint messages (barriers) being sent from sources to downstream operators, we will need to implement those. These may or may not depend on Watermarks task.
4. Implement a CheckpointCoordinator actor/logic which tracks state of checkpoints, health of the pipeline and triggers restores of the job's state when necessary. This may also be a part of a bigger actor/system tracking an overall health of the cluster

A good video to watch to understand possible implementations of checkpointing and trade-offs (checkpoint duration/frequency vs checkpoint memory footprint): https://www.youtube.com/watch?v=-ABPHMhrecQ
 TL;DR:

- _aligned checkpoinitng_: blocks all input streams until all checkpoint barriers are met, only then makes a state checkpoint. Long (in some cases - hours), but small memory footprint
- _unaligned checkpoinitng_: starts state checkpointing as soon as at least one barrier arrived, stores all of the incoming messages until all of the barriers arrived. Faster but requires more memory (GBs in some cases)
- There is also a thing called _incremental checkpoinitng_: a version of unaligned checkpoinitng, where messages are stored as soon as they arrive (async)
